### PR TITLE
Disallow subqueries without a relation in the range table list

### PR DESCRIPTION
**Pull request by onderkalaci**
This commit disallows subqueries without relation in the range table
list. This commit is only applied for subquery pushdown and INSERT..
SELECT queries. In other words, we do not add this limitation for single
table re-partition subqueries.

The reasoning behind this limitation is that if we allow pushing down
such queries, the result would include (shardCount * expectedResults)
where in a non distributed world the result would be (expectedResult)
only.

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -838,6 +838,15 @@ MultiTaskRouterSelectQuerySupported(Query *query)
 
 		Assert(subquery->commandType == CMD_SELECT);
 
+		/* pushing down rtes without relations yields (shardCount * expectedRows) */
+		if (subquery->rtable == NIL)
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
```
Comment (metdos) [inline src/backend/distributed/planner/multi_router_planner.c]: Is this a fix for INSERT ... SELECT queries? If yes, shall we do that separately?

File: src/test/regress/expected/multi_insert_select_non_pushable_queries.out [hunk 2]
```diff
@@ -660,3 +660,32 @@ FROM
...
+    user_id,
+    user_lastseen,
+    array_length(event_array, 1)
+FROM (
+    SELECT
+        user_id,
+        max(u.time) as user_lastseen,
+        array_agg(event_type ORDER BY u.time) AS event_array
+    FROM (
+
```
Comment (metdos) [inline src/test/regress/expected/multi_insert_select_non_pushable_queries.out]: Empty line. Also, we can fix query format a bit.

**Comment by metdos** [comment]
Adding back disappeared comment:

This looks good. I have two comments;

- We also change the point where we error out for INSERT ... SELECT queries. As long as I tested, we already detect this problem at another place and error out with a different message. Do we want to make this change here? (I prefer this approach to old one)
- Could you go over regression tests, and fix some indentation issues?

File: src/backend/distributed/planner/multi_router_planner.c (hunk 1 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c]: I moved this to a separate commit. Does that work?

File: src/test/regress/expected/multi_insert_select_non_pushable_queries.out [hunk 3]
```diff
@@ -660,3 +660,31 @@ FROM
...
+FROM (
+    SELECT
+        user_id,
+        max(u.time) as user_lastseen,
+        array_agg(event_type ORDER BY u.time) AS event_array
+    FROM (
+        SELECT user_id, time, value_3 as val_3
+        FROM users_table
+        WHERE
+        user_id >= 10 AND
```
Comment (metdos) [inline src/test/regress/expected/multi_insert_select_non_pushable_queries.out]: why user_id's are on different lines but value_1's are on same line :)

File: src/test/regress/sql/multi_insert_select_non_pushable_queries.sql [hunk 4]
```diff
@@ -649,3 +649,31 @@ FROM
...
+FROM (
+    SELECT
+        user_id,
+        max(u.time) as user_lastseen,
+        array_agg(event_type ORDER BY u.time) AS event_array
+    FROM (
+        SELECT user_id, time, value_3 as val_3
+        FROM users_table
+        WHERE
+        user_id >= 10 AND
```
Comment (metdos) [inline src/test/regress/sql/multi_insert_select_non_pushable_queries.sql]: why user_id's are on different lines but value_1's are on same line :)

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql [hunk 5]
```diff
@@ -1285,6 +1285,59 @@ SELECT * FROM run_command_on_workers('DROP FUNCTION array_index(ANYARRAY, ANYELE
 ORDER BY 1,2;
 DROP FUNCTION array_index(ANYARRAY, ANYELEMENT);
 
+-- a not supported query due to constant range table entry
+SELECT count(*) as subquery_count
+FROM (
+  SELECT 
+      user_id
+    FROM
+    users_table
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: I couldn't see any pattern on above 3 lines.

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql [hunk 6]
```diff
@@ -1285,6 +1285,59 @@ SELECT * FROM run_command_on_workers('DROP FUNCTION array_index(ANYARRAY, ANYELE
...
+  WHERE
+    (value_1 = '5' OR value_1 = '13')
+  GROUP BY user_id
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: we have user_id on the same line with GROUP BY, but for SELECT we use the next line.

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql [hunk 7]
```diff
@@ -1285,6 +1285,59 @@ SELECT * FROM run_command_on_workers('DROP FUNCTION array_index(ANYARRAY, ANYELE
...
+  SELECT
+    1 as user_id
+  ) AS b 
+  ON a.user_id = b.user_id 
+WHERE b.user_id IS NULL
+GROUP BY a.user_id;
+
+-- this is slightly different, we use RTE_VALUEs here
+SELECT Count(*) AS subquery_count 
+FROM   (SELECT user_id
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: More than one space.

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql [hunk 8]
```diff
@@ -1285,6 +1285,59 @@ SELECT * FROM run_command_on_workers('DROP FUNCTION array_index(ANYARRAY, ANYELE
...
+  ON a.user_id = b.user_id 
+WHERE b.user_id IS NULL
+GROUP BY a.user_id;
+
+-- this is slightly different, we use RTE_VALUEs here
+SELECT Count(*) AS subquery_count 
+FROM   (SELECT user_id 
+        FROM   users_table 
+        WHERE  ( value_1 = '5' 
+                  OR value_1 = '13' )
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: Looking to pattern here ` OR value_1 = '13' ) ` should continue on the line above.

File: src/test/regress/sql/multi_subquery_complex_queries.sql [hunk 9]
```diff
@@ -1786,5 +1786,91 @@ FROM     (
...
+  ( SELECT *, random()
+   FROM
+     ( SELECT "t"."user_id",
+              "t"."time",
+              unnest("t"."collected_events") AS "event_types"
+      FROM
+        ( SELECT "t1"."user_id",
+                 min("t1"."time") AS "time",
+                 array_agg(("t1"."event")
+                           ORDER BY TIME ASC, event DESC) AS collected_events
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_complex_queries.sql]: ORDER BY TIME indented too much.

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql (hunk 6 above)
Comment (onderkalaci) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: I don't think we have a convention for this in our tests?

File: src/test/regress/sql/multi_subquery_complex_queries.sql (hunk 9 above)
Comment (onderkalaci) [inline src/test/regress/sql/multi_subquery_complex_queries.sql]: But that `ORDER BY` belongs to `array_agg`, so should be fine?

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql (hunk 6 above)
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: I think we don't have a formal one, but having symmetry and consistent spacing makes it easier to read and understand queries. It doesn't need to be strict as much as our coding conventions.

File: src/test/regress/sql/multi_subquery_complex_queries.sql (hunk 9 above)
Comment (metdos) [inline src/test/regress/sql/multi_subquery_complex_queries.sql]: Ohh, maybe we should put them in one line :)

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql (hunk 5 above)
Comment (onderkalaci) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: Should be fixed now?

File: src/test/regress/sql/multi_subquery_behavioral_analytics.sql [hunk 10]
```diff
@@ -1285,6 +1285,66 @@ SELECT * FROM run_command_on_workers('DROP FUNCTION array_index(ANYARRAY, ANYELE
...
+GROUP BY a.user_id;
+
+-- same with INNER JOIN
+SELECT count(*) as subquery_count
+FROM (
+  SELECT 
+      user_id
+    FROM
+    users_table
+  WHERE
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_behavioral_analytics.sql]: Unaligned lines.

File: src/test/regress/sql/multi_subquery_union.sql [hunk 11]
```diff
@@ -605,5 +605,73 @@ FROM
...
+  (SELECT (random() * 100)::int)
+) b;
+
+-- we don't support subqueries without relations
+SELECT 
+  user_id, value_3
+FROM 
+(
+  (SELECT value_3, user_id FROM events_table where event_type IN (1, 2, 3, 4, 5))
+    UNION ALL
```
Comment (metdos) [inline src/test/regress/sql/multi_subquery_union.sql]: first union indented but not others
